### ISSUE TITLE
OpenSSL 3.0.1: FIPS provider not available

### ISSUE THREAD

**Issue by biao-billy** [openssl/openssl#17883]
In migrating my OpenSSL 1.0.2 FIPS app to OpenSSL 3.0.1, I am having difficulty in loading the FIPS module. I am following closely the **"Programmatically loading the FIPS module (default library context)"** section of the [fips module guide](https://www.openssl.org/docs/manmaster/man7/fips_module.html) doc.

With the following changes, FIPS module failed to load. In fact, `OSSL_PROVIDER_available(NULL, "fips")` returns 0.

Can someone please review my changes and see what might still be missing?

**Config/Installation changes:**

1. `enable-fips`, `--api=1.1.1`, and `--libdir=lib` were added to `config` (latter 2 for app compat)
2. After build & install, `openssl-3.0.1/providers/fips.so`, `openssl-3.0.1/providers/fipsmodule.cnf` and `openssl-3.0.1/apps/openssl.cnf` were copied from the build VM to the app VM.

**App changes:**

1. Edited the `openssl.cnf` to have the settings described in the fips module guide.
2. Pointed app to the openssl config location: `OPENSSL_config("/opt/uno/etc/openssl.cnf")`
3. Added the following loading code:
```
            const char* default_search_path = "/opt/uno/lib/crypto";
            OSSL_PROVIDER_set_default_search_path(NULL, default_search_path);

            if (!OSSL_PROVIDER_available(NULL, "fips")) {
               SvcLog(LEVEL_ERROR, "FIPS provider not available");
               return false;
            }

            fips = OSSL_PROVIDER_load(NULL, "fips");
            if (fips == NULL) {
                SvcLog(LEVEL_ERROR,"Failed to load FIPS provider\n");
                return false;
            }
```
**Key file locations in app:**
```
[root@ven-rhel lib]# ls /opt/uno/lib/crypto/fips.so
/opt/uno/lib/crypto/fips.so

[root@ven-rhel etc]# ls /opt/uno/etc/openssl.cnf /opt/uno/etc/fipsmodule.cnf
/opt/uno/etc/fipsmodule.cnf  /opt/uno/etc/openssl.cnf
```
 **Selected content from openssl.cnf**
```
openssl_conf = openssl_init

.include fipsmodule.cnf

[openssl_init]
providers = provider_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1
```

**fipsmodule.cnf content**
```
[root@ven-rhel etc]# cat fipsmodule.cnf
[fips_sect]
# activate = 1
conditional-errors = 1
security-checks = 1
module-mac = 67:48:52:5C:46:50:3C:D9:FD:73:DD:77:C0:51:DB:1C:32:48:8D:73:DA:CF:55:B6:38:C9:40:C1:54:7B:B6:D1
```

**Comment by paulidale** [comment]
You need to load the FIPS provider either via configuration or explicitly.  Please read the [migration guide](https://www.openssl.org/docs/manmaster/man7/migration_guide.html).

**Comment by biao-billy** [comment]
@paulidale Is this wrong? It's referenced from the migration guild. How should I interpret it?

> Applications may choose to load the FIPS provider explicitly rather than relying on config to do this. **The config file is still necessary in order to hold the FIPS module config data** (such as its self test status and integrity data)

**Comment by paulidale** [comment]
The FIPS provider needs the bits `openssl fipsinstall` creates.  This is unconditional.

Your `fipsmodule.cnf` looks reasonable apart from the commented out activate line.

**Comment by mattcaswell** [comment]
Are you seeing any error message on the OpenSSL error stack, e.g. via `ERR_print_errors_fp(stderr)` or similar call?

**Comment by mattcaswell** [comment]
The code to load the FIPS provider looks ok to me at first glance.

**Comment by t8m** [comment]
Most probably the problem is with the relative path used in the `.include fipsmodule.cnf` directive. You need to use an absolute path or set OPENSSL_CONF_INCLUDE environment variable to the path where the fipsmodule.cnf is placed.

**Comment by t8m** [comment]
Otherwise the fipsmodule.cnf will be looked from the CWD of the application.

**Comment by paulidale** [comment]
Without running `openssl fipsinstall` you won't have the module checksum and the module won't load.  So yes, it must be run.
This lives in the configuration file `fipsmodule.cnf` so you need that too -- regardless of how the FIPS provider is loaded. It's easiest via config.

**Comment by t8m** [comment]
> I have a couple fundamental questions:
> 
>     1. Is FIPS module load by config file and by explicit API calls mutually exclusive or not. As I quoted from fips module guide above, the guide is pretty clear you need config file even if you load `fips.so` explicitly. Now what's not clear is for manual load, you need just `fipsmodule.cnf`? or you still need `openssl.cnf` where `fipsmodule.cnf` is referenced. @paulidale it would be great if you can clarify on that.

You need to somehow load the fipsmodule.cnf file. The easiest way to do that is to properly include it from the default openssl.cnf file.

>     2. Is `openssl fipsinstall` build target required? I don't know why that's the case based on fips module guide. @paulidale can you please elaborate on why including this **special purpose** build target is unconditional?

`make install` in a build with enable-fips should create the fipsmodule.cnf with the module checksum. Doesn't it for you?
 
> 
> > ... using the enable-fips option. If it is enabled, the FIPS provider gets built and installed in addition to the other standard providers. No separate installation procedure is necessary. There is however a dedicated install_fips make target, **which serves the special purpose of installing only the FIPS provider into an existing OpenSSL installation**.
> 
>     3. Must `fipsmodule.cnf` be placed in the same directory as `fips.so` in the app?

The fipsmodule.cnf can be placed anywhere. The point is that it must be somehow loaded which is not done automatically.

> @t8m, the sample `openssl.cnf` uses non- absolute path. I tried `.include ./fipsmodule.cnf` to no avail.

`./fipsmodule.cnf` is not an absolute path. You need to use absolute path or set `OPENSSL_CONF_INCLUDE` environment variable, otherwise the file is looked up in the current working directory of the process that loads the openssl library.

> ```
> # For FIPS
> # Optionally include a file that is generated by the OpenSSL fipsinstall
> # application. This file contains configuration data required by the OpenSSL
> # fips provider. It contains a named section e.g. [fips_sect] which is
> # referenced from the [provider_sect] below.
> # Refer to the OpenSSL security policy for more information.
> # .include fipsmodule.cnf
> ```

Yeah, we should adjust the sample openssl.cnf to not use the relative path.

**Comment by biao-billy** [comment]
Ok I have an update: I started to get error callbacks (I don't know why it decided to show up).

What's interesting about the callback is that it started **before** I made any explicit OpenSSL calls. It looks as if loading of libcrypto (which is auto loaded by the app) is somehow causing OpenSSL to look for FIPS config? Both `libcrypto.so` and `fips.so` are in `/opt/uno/lib/crypto/` dir. My app doesn't tell the loader to load `fips.so`, only `libcrypto.so` and `libssl.so`.

Our installer does not copy `fipsmodule.cnf` to `/opt/uno/lib/crypto/`. I copied the file there and the error callback stopped. 

Any idea why this is happpening?

```
2022-03-16T08:10:18.998-07:00 INFO:: calling ERR_print_errors_cb.
2022-03-16T08:10:18.998-07:00 INFO:: OpenSSL: C077D462647F0000:error:80000002:system library:process_include:No such file or directory:crypto/conf/conf_def.c:805:calling stat(/opt/uno/lib/crypto/fipsmodule.cnf)
2022-03-16T08:10:18.998-07:00 INFO:: OpenSSL: C077D462647F0000:error:07800069:common libcrypto routines:provider_conf_load:provider section error:crypto/provider_conf.c:156:section=fips_sect not found
2022-03-16T08:10:18.998-07:00 INFO:: OpenSSL: C077D462647F0000:error:0700006D:configuration file routines:module_run:module initialization error:crypto/conf/conf_mod.c:243:module=providers, value=provider_sect retcode=-1
2022-03-16T08:10:18.999-07:00 INFO:: calling OPENSSL_config(/opt/uno/etc/openssl.cnf)
2022-03-16T08:10:18.999-07:00 INFO:: calling OSSL_PROVIDER_set_default_search_path w/  /opt/uno/lib/crypto
2022-03-16T08:10:18.999-07:00 INFO:: calling OSSL_PROVIDER_load w/ "fips"
2022-03-16T08:10:18.999-07:00 ERROR:: Failed to load FIPS provider
```

**Comment by biao-billy** [comment]
**Update**: I found an email thread archived discussing a very similar issue as mine. So I reached out to Jason Schultz who was trying to get FIPS provider to load. He told me he eventually got the FIPS provider to load - with procedure described in this [email](https://www.mail-archive.com/openssl-users@openssl.org/msg90210.html). He also stated that his code only works on the machine that builds OpenSSL 3.0.

So I added the missing call, `OSSL_LIB_CTX_load_config`, and tried. It didn't work. Then I thought I should try my app on the build VM. And just as what he mentioned, it worked!

Obviously, that's not a solution.

So I figured there must be some states, other than `fips.so` and `fipsmodule.cnf` that are left behind in the build machine. And what could those be?

Anyone knows or have a hypothesis?

**Comment by t8m** [comment]
You could try to use `strace` to find out what files is the application trying to load.

**Comment by biao-billy** [comment]
Hi all, I am finally able to get FIPS provider to load in my app VM. The last issue was my script use `cp` instead of `cp -p` to copy the `fips.so`. Evidently the change of timestamp and/or ownership caused integrity check to fail.

Thanks everyone. Special thanks to @t8m !

Billy

**Comment by biao-billy** [comment]
@t8m You are right file attributes didn't invalidate the integrity. Rather, my script unconditionally striped symbols from all libraries, including `fips.so`, which changed the file content. That's the real cause for failing integrity check.

Thanks again for your assistance.

**Comment by raghupcr** [comment]
@biao-billy

 hi, I could see this issue is closed and it is resolved.

- I could see from the above discussion that 

=============================
**- So I added the missing call, OSSL_LIB_CTX_load_config, and tried. It didn't work. Then I thought I should try my app on the build VM. And just as what he mentioned, it worked!

Obviously, that's not a solution.**
=====================================

so it worked on Build VM but on the app VM it failed is what i understood.

Later in the end we saw cp -p is the solution?

one doubt is did you run the command openssl fipsinstall (openssl-3.0.1/dist/bin/openssl fipsinstall command) on both build VM and app VM?

Is it done on Build VM alone and later copied the files (fipmodule.cnf and fips.so) then it worked for you?

I know this issue is in closed state but just wanted to get some infor which will be helpful for our application
Thanks,

**Comment by paulidale** [comment]
Something that hasn't been noted yet: copying `fipsmodule.cnf` from the build VM to another VM invalidates FIPS.  The security policy is quite explicit:

> Note: The Module shall have the self-tests run, and the Module config file output generated on each
platform where it is intended to be used. The Module config file output data shall not be copied from
one machine to another.

You **must** run the fipsinstall process on each operational environment.   Before anyone asks, a different VM is a different operational environment.

As for the question being asked, the problem was the stripping of symbols without regenerating the integrity checksum.  The `cp -p` did fix this, but so would a normal `cp`.

**Comment by raghupcr** [comment]
@paulidale 
Good Morning.

Thanks Dr. Paul.

### Note: The Module shall have the self-tests run, and the Module config file output generated on each
platform where it is intended to be used. The Module config file output data shall not be copied from
one machine to another.
- Yes I completely agree with you.

If we copy both fipmodule.cnf and fips.so/dll together to a different VM, will that solve the issue? Still, I agree that it's not FIPS validated. we have this use case so we were trying POC and if we copy both fips enablement was working so when i saw the biao-billy use case it is exactly same copying to different machine and  enabling fips.

But still i m with you Dr. Paul that "The Module config file output data shall not be copied from
one machine to another". we are doing POC

**Comment by paulidale** [comment]
> If we copy both fipmodule.cnf and fips.so/dll together to a different VM, will that solve the issue? 

No.  Fipsinstall **must** be run on each VM.  Anything else is not FIPS.

Yes, it produces _exactly_ the same output for each but that' not the point.  The primary reason for this little piece of craziness is because the self tests are run once at installation time rather than on every load.  Without using fipsinstall, you haven't run the self tests properly.

**Comment by raghupcr** [comment]
@paulidale 
Thanks you Dr Paul.

Agreed. Just for the knowledge to share with the teams in Dell to get more details.

**No. Fipsinstall must be run on each VM. Anything else is not FIPS.
Without using fipsinstall, you haven't run the self tests properly.
==============**

only when we do fipsinstall on another machine then only self tests are run and not on provider load, fips enablement etc.
we want at the application level and not at the system level, then how to run fipsinstall from the application side?

should we carry/ install OpenSSL 3.0.x on each VM is this the pre-requisite for enabling FIPS mode (example for1.0.2 - fips_mode_set(1)) on each machine? If so should we bundle complete openssl files in our bundle and run fipsinstall on the remote machine?

**Yes, it produces exactly the same output for each but that' not the point.
=============**
- so if we dont do that fipsinstall then the results will vary at run time or will be there be any consequences later?

Thanks

**Comment by paulidale** [comment]
Answering the questions inorder:

1. You run fipsinstall via the command line utility: `openssl fipsinstall ...`.  There is no other way.
2. You need to run the openssl command line utility on each VM.  The easiest way is to install the openssl command line utility but this isn't mandatory.  It could be run over a network.  It can also be deleted after running.
3. If you don't run fipsinstall and, instead, copy the fip configuration, everything will work fine.  At least until you are audited and it is discovered that your application is not FIPS compliant.  This is a very bad position to end up in and is the kind of thing the board of directors will hear about and action.  I wouldn't take the risk.

**Comment by raghupcr** [comment]
Sure Dr Paul we will take your inputs

- You run fipsinstall via the command line utility: openssl fipsinstall .... There is no other way.
   _OK Sure_
- You need to run the openssl command line utility on each VM. The easiest way is to install the openssl command line utility but this isn't mandatory. It could be run over a network. It can also be deleted after running.
  _So we need to install OpenSSL 3.0.x on the VM then we can call openssl fipsinstall on that machine as we cant run over the network as these are customer machines. we need to explicitly ask the customers to move to OpenSSL 3.0.x so that we can use that libraries to enable FIPS compliance_
- If you don't run fipsinstall and, instead, copy the fip configuration, everything will work fine. At least until you are audited and it is discovered that your application is not FIPS compliant. This is a very bad position to end up in and is the kind of thing the board of directors will hear about and action. I wouldn't take the risk.
_Ok Sure we need to check with our security team consultants and take the necessary steps_

Thanks for the inputs

**Comment by raghupcr** [comment]
@paulidale 

Hi Dr Paul,

- Do we need to have the requirement to the customer where ever our application is supposed to run to have the OpenSSL 3.0.x as the prerequisite?
- Or Carry complete OpenSSL required files
-  Later run the openssl fipsinstall

Is our understanding correct?

Thanks

**Comment by paulidale** [comment]
I don't understand the question.  I'm also not sure how I can explain it again any better than I have.

Run fipsinstall before using the FIPS provider on each and every machine.  How you do this is up to you and your individual setup.

**Comment by raghupcr** [comment]
I don't understand the question
- Will try to make it bit clear as i have mentioned we are running on another product/machine and we bundle all the files which are required for our application to run and install on the product's machine. So to run fipsinstall command whether we need to carry all the required files (openssl.exe, etc) or do we need to ask the product only to install OpenSSL 3.0.x and run fipsinstall which is applicable for the complete system
-

**Comment by paulidale** [comment]
Both would work, the decision is up to you, the owners of the other machine and the complexities, advantages and disadvantages of each approach.

The OpenSSL project, cannot assist in this decision, we don't know the individual setup in anything like sufficient detail.

**Comment by raghupcr** [comment]
@paulidale 

Sure Thanks for the information
